onnxruntime/core/providers/cuda/reduction/reduction_ops.cc

```diff
@@ -349,73 +340,65 @@ static Status PrepareForReduce(OpKernelContext* ctx,
-  CUDA_RETURN_IF_ERROR(cudaMemset(Y->MutableDataRaw(), 0, Y->SizeInBytes()));
-  *y_pp = Y;
-
   // CUDNN requires at least 3D input, so pad 1s if needed
-  input_dims_cudnn = input_dims;
-  output_dims_cudnn = output_dims;
+  prepare_reduce_metadata.input_dims_cudnn = input_dims;
+  prepare_reduce_metadata.output_dims_cudnn = prepare_reduce_metadata.output_dims;
   if (rank < 3) {
     std::vector<int64_t> pads(3 - rank, 1);
-    input_dims_cudnn.insert(input_dims_cudnn.end(), pads.begin(), pads.end());
-    output_dims_cudnn.insert(output_dims_cudnn.end(), pads.begin(), pads.end());
+    prepare_reduce_metadata.input_dims_cudnn.insert(prepare_reduce_metadata.input_dims_cudnn.end(), pads.begin(), pads.end());
+    prepare_reduce_metadata.output_dims_cudnn.insert(prepare_reduce_metadata.output_dims_cudnn.end(), pads.begin(), pads.end());
   }
 
-  output_count = Y->Shape().Size();
+  prepare_reduce_metadata.output_count = TensorShape(prepare_reduce_metadata.output_dims).Size();
 
-  if (rank == 0) {
-    rank = 1;
+  if (prepare_reduce_metadata.rank == 0) {
+    prepare_reduce_metadata.rank = 1;
   }
 
   return Status::OK();
 }
 
-template <bool allow_multi_axes>
+// `input_shape_override` is the input shape for compute purposes (if provided)
 template <typename T, cudnnReduceTensorIndices_t ReduceTensorIndices>
-Status ReduceKernel<allow_multi_axes>::ComputeImpl(OpKernelContext* ctx, cudnnReduceTensorOp_t cudnn_reduce_op) const {
+static Status ReduceComputeCore(CUDAExecutionProvider& cuda_ep, const Tensor& input, PrepareReduceMetadata& prepare_reduce_metadata,
+                                /*out*/ Tensor& output, cudnnReduceTensorOp_t cudnn_reduce_op,
+                                const std::vector<int64_t>& axes,
+                                bool calculate_log, bool calculate_sqt, bool log_sum_exp, bool fast_reduction,
+                                const TensorShape* input_shape_override = nullptr) {
   typedef typename ToCudaType<T>::MappedType CudaT;
-  const Tensor* X = nullptr;
-  Tensor* Y = nullptr;
-
-  int64_t input_count = 0;
-  int64_t output_count = 0;
-  std::vector<int64_t> output_dims;
-  std::vector<int64_t> input_dims_cudnn;
-  std::vector<int64_t> output_dims_cudnn;
-  int64_t rank = 0;
-  int64_t stride = 0;
-  ORT_RETURN_IF_ERROR(PrepareForReduce(ctx,
-                                       keepdims_,
-                                       axes_,
-                                       &X,
-                                       &Y,
-                                       input_count,
-                                       output_count,
-                                       output_dims,
-                                       input_dims_cudnn,
-                                       output_dims_cudnn,
-                                       rank, stride));
+  const Tensor* X = &input;
+  const TensorShape& input_shape = input_shape_override ? *input_shape_override : X->Shape();
+  Tensor* Y = &output;
+
```

Comment:
nit: Tensor& Y = output. Or just use 'output' directly.

Reply:
used input and output directly.